### PR TITLE
Improve handling of I/O requests and on-demand page allocation under memory pressure

### PR DESCRIPTION
**Pull request by francescolavra**
This change set improves handling of out-of-memory conditions so that no error is returned to userspace when there is reclaimable memory that can be used to satisfy a request from the user program.
More specifically, read and write requests in the page cache code can now trigger a "memory cleaning" operation (possibly suspending the current context if cached data needs to be flushed to disk) if a page allocation fails. Analogously, if an anonymous page cannot be allocated on-demand due to a shortage of available physical memory, the context that triggered a page fault is suspended waiting for cached data to be flushed to disk, so that the page allocation can be re-tried after reclaimable memory has been freed.
In addition, handling of faulting user memory in the getrandom() syscall has been improved so that a non-resolvable page fault (due to either an invalid memory range or an out-of-memory condition) results in an EFAULT return code instead of a kernel crash.

### CODE REVIEW

**Review by wjhun — APPROVED** [review]
LGTM

File: src/unix/syscall.c [hunk 1]
```diff
@@ -1133,39 +1133,44 @@ sysreturn creat(const char *pathname, int mode)
...
 #define GETRANDOM_MAX_BUFLEN (1ull << 20)
 
-static inline void fill_random(void *buf, u64 buflen)
+static inline boolean fill_random(void *buf, u64 buflen)
 {
+    context ctx = get_current_context(current_cpu());
+    if (context_set_err(ctx)) {
+        random_buffer_aborted();
+        return false;
+    }
```
[the remark this answers is not in the document]
Comment (francescolavra) [reply src/unix/syscall.c:1142]: Returning the bytes successfully copied seems more consistent with how we deal with other syscalls operating on user buffers, and also Linux now does that (the previous behavior has been changed in [this commit "random: allow partial reads if later user copies fail"](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/drivers/char/random.c?id=5209aed5137880fa229746cb521f715e55596460)).

File: src/kernel/pagecache.c [hunk 2]
```diff
@@ -753,7 +781,33 @@ closure_function(1, 3, void, pagecache_write_sg,
         pn->length = q.end;
 
     pagecache_unlock_node(pn);
-    apply(sh, STATUS_OK);
+    if (err_msg) {
+#ifdef KERNEL
+        if (!mem_cleaned || (pi != r.start)) {
```
[the remark this answers is not in the document]
Comment (francescolavra) [reply src/kernel/pagecache.c:786]: Yes, making some progress after a previous mm_service() call means that more pages might have been allocated to service this request, which would then trigger some more "memory cleaning" in the next_mm_service() call.

**Comment by francescolavra** [comment]
I added a commit to address an additional issue discovered by @rinor (https://github.com/nanovms/nanos/issues/1845#issuecomment-1655358623).

**Comment by rinor** [comment]
Is there any way we can help to speed up the merge/rebase of this pr? I understand the time constraints and maybe other higher priority work, so no rush, just asking. Thank you.

**Comment by francescolavra** [comment]
I rebased this PR on the master branch. It still needs to be reviewed before merging, more specifically the last 6 commits, which have been added after the last review.

**Comment by rinor** [comment]
Thank you!

File: src/runtime/tuple.c [hunk 3]
```diff
@@ -185,6 +187,32 @@ tuple allocate_tuple(void)
     return tag(allocate_table(theap, key_from_symbol, pointer_equal), tag_table_tuple);
 }
 
+closure_function(1, 2, boolean, clone_tuple_each,
+                 tuple, clone,
+                 value, s, value, v)
+{
+    if (is_tuple(v))
+        v = clone_tuple(v);
+    else
```
Comment (sanderssj) [inline src/runtime/tuple.c:196]: This needs to be updated handle the new tag types.

**Review by sanderssj — COMMENTED** [review]
LGTM other than dealing with the new tag types.

File: src/runtime/tuple.c (hunk 3 above)
Comment (francescolavra) [reply src/runtime/tuple.c]: OK, the clone_tuple_each closure now checks whether a value is immediate (and also whether it's `null_value`), and handles the value based on its tag.

**Pull request merged**